benchmark/C/Savina/src/micro/Counting.lf

```diff
@@ -0,0 +1,151 @@
+ # See https://nedbatchelder.com/code/cog/
+ #
+ # All instructions for code  generation are in-lined in comments
+ # like this one. With that you can use this file as a normal source file
+ # but also to generate code.
+ #
+ # To change the generated code in-line within this file run:
+ # $ python -m cog -r this-file.lf
+ # To generate a new file from this file stripping the generator code in the process run:
+ # $ python -m cog -d -o output-file.lf this-file.lf
+ #
+ # Use the command line option -D to specify generator parameters, for example:
+ # $ python -m cog -r -D parameter=100 this-file.lf
+ #
+ # Generator parameters used in this file:
+ # -D countTo=1000000
+ #
+ ]]] */
+ // [[[end]]]
+
+target C {
+    tracing: true
+};
+
+
+reactor ProducerReactor(countTo:int(1000000)) {
+
+
+    state i:int(0); // current counter value
+
+    input inStartCounting:bool;
+
+	logical action outFinishedCounting;
+    output outResetCounter:bool;
+    output outIncrement:bool;
+    output outRetrieveResult:bool;
+    input inResult:int;
+    logical action count;
+
+    reaction(outFinishedCounting){=
+        request_stop();
+    =}
+
+
+    reaction(inStartCounting) -> count, outResetCounter {=
+        // reset counter
+        self->i = 0;
+        SET(outResetCounter, true);
+
+        schedule(count,0);
+    =}
+
+    reaction(count) -> outIncrement, outRetrieveResult, count {=
+
+        if(self->i < self->countTo) {
+            SET(outIncrement,true);
+            self->i += 1;
+            schedule(count,0);
+        } else {
+            SET(outRetrieveResult,true);
+        }
+    =}
+
+    reaction(inResult) -> outFinishedCounting {=
+        if(inResult->value != self->countTo) {
+            printf("ProducerReactor expected %d, found: %d",self->countTo, inResult->value);
+        } else {
+            printf("SUCCESS! received: %d ", inResult->value);
+        }
+        schedule(outFinishedCounting, 0);
+    =}
+}
+
+
+reactor CountingReactor (numIterations:int(12)){
+
+    input inFinishedCounting:bool;
+    input inReset:bool;
+    input inIncrement:bool;
+    input inRetrieveResult:bool;
+    output outSendResult:int;
+    output outStartCounting:bool;
+
+    state count:int(0);
+    state iteration:int(0);
+
+    reaction(inFinishedCounting) -> outStartCounting{=
+        if(self->numIterations <= self->iteration)
+        {
+          request_stop();
+        }
+        else{
+          self->iteration += 1;
+          printf("Iteration: %d",self->iteration);
+          SET(outStartCounting,true);
+        }
+    =}
+
+    reaction(inReset) {=
+        self->count = 0;
+    =}
+
+    reaction(inIncrement) {=
+        self->count++;
+    =}
+
+    reaction(inRetrieveResult) -> outSendResult {=
+        SET(outSendResult,self->count);
+    =}
+}
+
+main reactor CountingBenchmarkN(numIterations:int(12), countTo:int(1000000)) {
+    
+    /* [[[cog
+    cog.outl(f"producer = new ProducerReactor(countTo = {countTo});")
```

Comment:
This is only a nitpick, but I think it is a bit cleaner to use cog to only change the default value of the `countTo` parameter above, and then pass this parameter along to the other reactors.

Reply:
Does this mean cog it at the top of the file and use it like a global variable?

Reply:
I am not sure what you mean by "at the top of the file" here. What I meant is that you can cog the main reactor declaration, such that you overwrite the default values of any parameters to the main reactor. Then you have one central place where the code is changed by cog and not multiple.

Also, this would help to bring the C and C++ code more in line. In C++, cog is not needed because parameters to the main reactor can be overwritten on the command line without recompiling.